swift/apple/FirezoneNetworkExtension/ConfigurationManager.swift

```diff
@@ -12,83 +12,95 @@ import CryptoKit
 class ConfigurationManager {
   static let shared = ConfigurationManager()
 
-  let encoder = PropertyListEncoder()
+  let userDictKey = "dev.firezone.configuration"
+  let managedDictKey = "com.apple.configuration.managed"
 
   private var userDefaults: UserDefaults
 
-  var authURL: URL? {
-    get { userDefaults.url(forKey: Configuration.Keys.authURL) }
-    set { userDefaults.set(newValue, forKey: Configuration.Keys.authURL) }
+  // We maintain a cache of the user dictionary to buffer against unnecessary reads from UserDefaults which
+  // can cause deadlocks in rare cases.
+  var userDict: [String: Any?]
+
+  var managedDict: [String: Any?] {
+    userDefaults.dictionary(forKey: managedDictKey) ?? [:]
   }
 
-  var apiURL: URL? {
-    get { userDefaults.url(forKey: Configuration.Keys.apiURL) }
-    set { userDefaults.set(newValue, forKey: Configuration.Keys.apiURL) }
+  private init() {
+    userDefaults = UserDefaults.standard
+    userDict = userDefaults.dictionary(forKey: userDictKey) ?? [:]
+
+    migrateFirezoneId()
+    Telemetry.firezoneId = userDict[Configuration.Keys.firezoneId] as? String
   }
 
-  var logFilter: String? {
-    get { userDefaults.string(forKey: Configuration.Keys.logFilter) }
-    set { userDefaults.set(newValue, forKey: Configuration.Keys.logFilter) }
+  func setAuthURL(_ authURL: URL) {
+    userDict[Configuration.Keys.authURL] = authURL.absoluteString
```

Comment:
Turns out the hang wasn't a deadlock - if you pass a dict with nested types that aren't one of the basic plist types, it would hang forever.

Fixed by converted URLs to/from Strings.

Reply:
I love Apple's APIs. And their documentation is even better (not).